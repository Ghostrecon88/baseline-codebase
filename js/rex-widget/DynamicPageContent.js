/**
 * @copyright 2016-present, Prometheus Research, LLC
 * @flow
 */

import * as React from "react";
import { VBox } from "react-stylesheet";
import { fetch } from "./fetch";
import * as rexui from "rex-ui";

type Props = {|
  content: React.Node,
  location: Location,
  onNavigation: (href: string) => boolean
|};

type State = {
  content: React.Node,
  updating: boolean
};

export default class DynamicPageContent extends React.Component<Props, State> {
  state: State = {
    content: this.props.content,
    updating: false
  };

  render() {
    let { location } = this.props;
    return (
      <VBox flexGrow={1} flexShrink={1} key={location.href}>
        {this.state.updating ? <rexui.PreloaderScreen /> : this.state.content}
      </VBox>
    );
  }

  componentWillReceiveProps({ location }: Props) {
    this.setState({ updating: true });
    fetch(location.href, {}, { useTransit: true }).then(
      this.onPageFetched.bind(null, location.href),
      this.onPageError
    );
  }

  onPageFetched = (href: string, page: { props: { content: React.Node } }) => {
    setTimeout(() => {
      this.setState({
        content: page.props.content,
        updating: false
      });
    }, 10);
  };

  onPageError = (error: Error) => {
    this.setState({ updating: false });
    console.error(error);
  };

  hijackLinkClick = (event: MouseEvent) => {
    if (!isMainClick(event)) {
      return;
    }
    let href = getHrefFromEventTarget(event.target);
    if (!href) {
      return;
    }
    if (this.props.onNavigation(href)) {
      event.preventDefault();
      event.stopPropagation();
    }
  };

  componentDidMount() {
    document.addEventListener("click", this.hijackLinkClick);
  }

  componentWillUnmount() {
    document.removeEventListener("click", this.hijackLinkClick);
  }
}

function isMainClick(event: MouseEvent) {
  return (
    event.button === 0 &&
    !event.metaKey &&
    !event.altKey &&
    // $FlowFixMe: ...
    !event.ctrlkey &&
    !event.shiftKey
  );
}

function getHrefFromEventTarget(element): null | string {
  do {
    if (element.tagName === "A" && element.href) {
      // $FlowFixMe: ...
      return element.href;
    }
    // $FlowFixMe: ...
    element = element.parentElement;
  } while (element);

  return null;
}
