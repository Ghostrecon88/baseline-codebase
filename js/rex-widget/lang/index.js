/**
 * @copyright 2016, Prometheus Research, LLC
 */

export {default as transferStaticProperties} from './transferStaticProperties';
export {default as isString} from './isString';
export {default as isArray} from './isArray';
export {default as delay} from './delay';
export {default as toSnakeCase} from 'to-snake-case';

export {default as emptyFunction} from './emptyFunction';
