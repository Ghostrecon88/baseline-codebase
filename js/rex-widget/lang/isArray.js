/**
 * @copyright 2015, Prometheus Research, LLC
 */

let toString = Object.prototype.toString;
const ARRAY_REP = '[object Array]';

export default function isArray(o) {
  return toString.call(o) === ARRAY_REP;
}
