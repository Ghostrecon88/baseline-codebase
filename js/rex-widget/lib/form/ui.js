/**
 * @copyright 2016-present, Prometheus Research, LLC
 */

import React from 'react';
import {style, VBox} from '@prometheusresearch/react-ui';

export let Hint = style('div', {
  base: {
    fontSize: '75%',
    textAlign: 'left',
    marginTop: '5px',
  },
});

export let FieldsetLabel = style('label', {
  base: {
    color: '#000',
    fontSize: '100%',
    fontWeight: 700,
  },
});

export let ErrorMessage = style('span', {
  base: {
    color: 'red',
    marginLeft: 3,
    marginRight: 3,
    display: 'inline-block',
  },
});

export let HeaderContainer = style(VBox, {
  base: {
    marginTop: 15,
    marginBottom: 15,
  },
});

export let RequiredSign = function() {
  return <ErrorMessage style={{width: 3}}>*</ErrorMessage>;
};

export let FieldsetHeader = function({label, hint, isRequired, ...props}) {
  if (!label && !hint) {
    return <noscript />;
  }
  return (
    <HeaderContainer {...props}>
      {label &&
        <FieldsetLabel>
          {label} {isRequired && <RequiredSign />}
        </FieldsetLabel>}
      {hint &&
        <Hint>
          {hint}
        </Hint>}
    </HeaderContainer>
  );
};
