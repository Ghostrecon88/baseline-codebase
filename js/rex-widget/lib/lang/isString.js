/**
 * @copyright 2015, Prometheus Research, LLC
 */

let toString = Object.prototype.toString;
const STRING_REP = '[object String]';

export default function isString(o) {
  return toString.call(o) === STRING_REP;
}
