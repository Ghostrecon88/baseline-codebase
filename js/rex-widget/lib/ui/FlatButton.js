/**
 * @copyright 2015, Prometheus Research, LLC
 */

import Button from './Button';
import {rgb, none} from '../CSS';

export default Button.style({
  raised: true,

  textWidth: 300,

  text: rgb(173),
  textHover: rgb(130),
  textFocus: rgb(154),
  textActive: rgb(173),
  textDisabled: '#dadada',

  background: rgb(255),
  backgroundHover: rgb(241),
  backgroundFocus: rgb(255),
  backgroundActive: rgb(241),
  backgroundDisabled: rgb(255),

  border: rgb(234),
  borderHover: rgb(224),
  borderFocus: rgb(204),
  borderActive: rgb(224),
  borderDisabled: '#ececec',

  shadow: none,
  shadowHover: none,
  shadowFocus: none,
  shadowActive: '#ddd',
  shadowDisabled: none,
});

