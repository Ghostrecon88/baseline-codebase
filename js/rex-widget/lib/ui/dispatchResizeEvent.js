/**
 * @copyright 2015, Prometheus Research, LLC
 */

export default function dispatchResizeEvent() {
  requestAnimationFrame(() => {
    let event = document.createEvent('UIEvents');
    event.initUIEvent('resize', true, false, window, 0);
    window.dispatchEvent(event);
  });
}
