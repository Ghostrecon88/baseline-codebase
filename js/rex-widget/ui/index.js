/**
 * @copyright 2015-present, Prometheus Research, LLC
 * @flow
 */

export { default as Focusable } from "./Focusable";

export { default as Icon } from "./Icon";

export { default as Label } from "./Label";
export { default as Divider } from "./Divider";
export { default as IFrame } from "./IFrame";

export { default as Notification } from "./Notification";
export { showNotification, removeNotification } from "./NotificationCenter";
export { default as TouchableArea } from "./TouchableArea";
export { default as Scroller } from "./Scroller";

import type { NotificationID } from "./NotificationCenter";
export type { NotificationID };
