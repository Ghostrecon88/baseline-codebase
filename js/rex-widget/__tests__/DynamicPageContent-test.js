/**
 * @copyright 2016-present, Prometheus Research, LLC
 * @flow
 */

import * as React from "react";
import * as ReactTesting from "react-testing-library";
import * as History from "rex-ui/History";
import * as TestHarness from "rex-ui/TestHarness";
import DynamicPageContent from "../DynamicPageContent";
import * as ui from "../ui";
import * as lang from "../lang";

function stubFetch(): { resolve: any => void, reject: Error => void } {
  let handle = { resolve: undefined, reject: undefined };
  let promise = {
    then(_onSuccess, _onError) {
      handle.resolve = _onSuccess;
      handle.reject = _onError;
      return promise;
    }
  };
  global.fetch = jest.fn().mockReturnValue(promise);
  return (handle: any);
}

describe("rex-widget", function() {
  describe("<DynamicPageContent />", function() {
    TestHarness.silenceConsoleError();

    let fetch = global.fetch;
    beforeEach(function() {
      fetch = global.fetch;
    });
    afterEach(function() {
      global.fetch = fetch;
    });

    afterEach(ReactTesting.cleanup);

    it("renders", async function() {
      let onNavigation = jest.fn();
      let rendered = ReactTesting.render(
        <DynamicPageContent
          content="HELLO"
          preloader="PRELOADER"
          location={History.createLocation("/")}
          onNavigation={onNavigation}
        />
      );
      expect(rendered.asFragment()).toMatchSnapshot();
    });

    it("re-fetches content on new location", async function() {
      let fetch = stubFetch();

      let onNavigation = jest.fn();
      let rendered = ReactTesting.render(
        <DynamicPageContent
          content="HELLO"
          preloader="PRELOADER"
          location={History.createLocation("/")}
          onNavigation={onNavigation}
        />
      );
      expect(rendered.asFragment()).toMatchSnapshot();

      rendered.rerender(
        <DynamicPageContent
          content="HELLO"
          preloader="PRELOADER"
          location={History.createLocation("/next")}
          onNavigation={onNavigation}
        />
      );
      expect(rendered.asFragment()).toMatchSnapshot();

      fetch.resolve({ props: { content: "NEXT HELLO" } });

      await lang.delay(20);

      expect(rendered.asFragment()).toMatchSnapshot();
    });

    it("re-fetches content on new location (error)", async function() {
      let fetch = stubFetch();

      let onNavigation = jest.fn();
      let rendered = ReactTesting.render(
        <DynamicPageContent
          content="HELLO"
          preloader="PRELOADER"
          location={History.createLocation("/")}
          onNavigation={onNavigation}
        />
      );
      expect(rendered.asFragment()).toMatchSnapshot();

      rendered.rerender(
        <DynamicPageContent
          content="HELLO"
          preloader="PRELOADER"
          location={History.createLocation("/next")}
          onNavigation={onNavigation}
        />
      );
      expect(rendered.asFragment()).toMatchSnapshot();

      fetch.reject(new Error("oops"));

      await lang.delay(20);

      expect(rendered.asFragment()).toMatchSnapshot();
    });
  });
});
