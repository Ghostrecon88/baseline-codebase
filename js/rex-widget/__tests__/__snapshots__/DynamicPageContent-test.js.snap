// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rex-widget <DynamicPageContent /> re-fetches content on new location (error) 1`] = `
<DocumentFragment>
  <div
    class="[object Object] [object Object] [object Object]"
  >
    HELLO
  </div>
</DocumentFragment>
`;

exports[`rex-widget <DynamicPageContent /> re-fetches content on new location (error) 2`] = `
<DocumentFragment>
  <div
    class="[object Object] [object Object] [object Object]"
  >
    PRELOADER
  </div>
</DocumentFragment>
`;

exports[`rex-widget <DynamicPageContent /> re-fetches content on new location (error) 3`] = `
<DocumentFragment>
  <div
    class="[object Object] [object Object] [object Object]"
  >
    HELLO
  </div>
</DocumentFragment>
`;

exports[`rex-widget <DynamicPageContent /> re-fetches content on new location 1`] = `
<DocumentFragment>
  <div
    class="[object Object] [object Object] [object Object]"
  >
    HELLO
  </div>
</DocumentFragment>
`;

exports[`rex-widget <DynamicPageContent /> re-fetches content on new location 2`] = `
<DocumentFragment>
  <div
    class="[object Object] [object Object] [object Object]"
  >
    PRELOADER
  </div>
</DocumentFragment>
`;

exports[`rex-widget <DynamicPageContent /> re-fetches content on new location 3`] = `
<DocumentFragment>
  <div
    class="[object Object] [object Object] [object Object]"
  >
    NEXT HELLO
  </div>
</DocumentFragment>
`;

exports[`rex-widget <DynamicPageContent /> renders 1`] = `
<DocumentFragment>
  <div
    class="[object Object] [object Object] [object Object]"
  >
    HELLO
  </div>
</DocumentFragment>
`;
